So I wrote a test for this because it seems it was missing and that was a good idea.

```text
- (time.Time) 2020-01-23 07:58:28.522377 +0100 CET m=+0.002462155,
+ (time.Time) 0001-01-01 00:00:00 +0000 UTC,
  (string) (len=11) "this is foo",
- (time.Time) 2020-01-23 07:58:28.522377 +0100 CET m=+0.002462155
+ (time.Time) 0001-01-01 00:00:00 +0000 UTC
```

Since I got the empty value I tried to debug what we're setting and what the field value ended up being:

```go
if field.IsBlank {
    _ = field.Set(bulkNow)

    fmt.Println(bulkNow)
    fmt.Println(field.Field.Interface())
}
```

This ended up giving me:

```text
2020-01-23 07:58:28.52324 +0100 CET m=+0.003324871
0001-01-01 00:00:00 +0000 UTC
```

So I thought, let's error check this:

```go
if err := field.Set(bulkNow); err != nil {
    panic(err)
}
```

And lo and behold, this gives me:

```text
panic: using unaddressable value [recovered]
        panic: using unaddressable value
```

I don't really have the time to dig into this now but would be nice to adress this before merging!

--- a/gorm_bulk.go
+++ b/gorm_bulk.go
@@ -6,17 +6,9 @@ import (
 	"reflect"
 	"sort"
 	"strings"
-	"time"
 
 	"github.com/jinzhu/gorm"
 )
-
-// bulkNow holds a global now state and will be used for each records as
-// CreatedAt and UpdatedAt value if they're empty. This value will be set to the
-// value from gorm.NowFunc() in scopeFromObjects to ensure all objects get the
-// same value.
-// nolint: gochecknoglobals
-var bulkNow time.Time
 
 // BulkInsert will call BulkExec with the default InsertFunc.
 func BulkInsert(db *gorm.DB, objects []interface{}) error {
@@ -94,14 +86,8 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 		placeholders      []string
 		groups            []string
 		scope             = db.NewScope(objects[0])
+		bulkNow           = gorm.NowFunc()
 	)
-
-	// Ensure we set the correct time and reset it after we're done.
-	bulkNow = gorm.NowFunc()
-
-	defer func() {
-		bulkNow = time.Time{}
-	}()
 
 	// Get a map of the first element to calculate field names and number of
 	// placeholders.
@@ -139,7 +125,18 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 		}
 
 		for _, key := range columnNames {
-			objectScope.AddToVars(row[key])
+			field := row[key]
+			value := field.Field.Interface()
+
+			switch field.Struct.Name {
+			// Column CreatedAt and UpdatedAt with zero value will be set to same time
+			case "CreatedAt", "UpdatedAt":
+				if field.IsBlank {
+					value = bulkNow
+				}
+			}
+
+			objectScope.AddToVars(value)
 		}
 
 		groups = append(
@@ -163,13 +160,10 @@ func scopeFromObjects(db *gorm.DB, objects []interface{}, execFunc ExecFunc) (*g
 //  * Fields marked to be ignored - Will be left out
 //  * Fields named ID with auto increment - Will be left out
 //  * Fields named ID set as primary key with blank value - Will be left out
-//  * Fields named CreatedAt or UpdatedAt with blank values - Will be set to
-//  gorm.NowFunc() value
 //  * Blank fields with default value - Will be set to the default value
-func ObjectToMap(object interface{}) (map[string]interface{}, error) {
+func ObjectToMap(object interface{}) (map[string]*gorm.Field, error) {
 	var (
-		attributes = map[string]interface{}{}
-		now        = bulkNow
+		attributes = map[string]*gorm.Field{}
 	)
 
 	// De-reference pointers (and it's values)
@@ -181,10 +175,6 @@ func ObjectToMap(object interface{}) (map[string]interface{}, error) {
 
 	if rv.Kind() != reflect.Struct {
 		return nil, errors.New("value must be kind of Struct")
-	}
-
-	if now.IsZero() {
-		now = gorm.NowFunc()
 	}
 
 	for _, field := range (&gorm.Scope{Value: object}).Fields() {
@@ -227,14 +217,7 @@ func ObjectToMap(object interface{}) (map[string]interface{}, error) {
 			}
 		}
 
-		if field.Struct.Name == "CreatedAt" || field.Struct.Name == "UpdatedAt" {
-			if field.IsBlank {
-				attributes[field.DBName] = now
-				continue
-			}
-		}
-
-		attributes[field.DBName] = field.Field.Interface()
+		attributes[field.DBName] = field
 	}
 
 	return attributes, nil

--- a/gorm_bulk_test.go
+++ b/gorm_bulk_test.go
@@ -430,3 +430,103 @@ func Test_columnOrder(t *testing.T) {
 		})
 	}
 }
+
+func Test_Timestamps(t *testing.T) {
+	db, _, err := sqlmock.New()
+	require.NoError(t, err)
+
+	gdb, err := gorm.Open("mysql", db)
+	require.NoError(t, err)
+
+	pastDate := time.Date(1985, 1, 1, 0, 0, 0, 0, time.UTC)
+	testStartedAt := time.Now()
+
+	cases := []struct {
+		description string
+		slice       []interface{}
+		expectedSQL string
+		validFunc   func(t *testing.T, vars []interface{})
+	}{
+		{
+			description: "set dates will persist",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo:       "this is foo",
+					CreatedAt: pastDate,
+					UpdatedAt: pastDate,
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+				assert.Equal(t, pastDate, createdAt)
+				assert.Equal(t, pastDate, updatedAt)
+			},
+		},
+		{
+			description: "empty time pointers will be set to now func",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt *time.Time
+					UpdatedAt *time.Time
+				}{
+					Foo: "this is foo",
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+
+				createdAtTime, ok := createdAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, createdAtTime.After(testStartedAt))
+
+				updatedAtTime, ok := updatedAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, updatedAtTime.After(testStartedAt))
+			},
+		},
+		{
+			description: "empty time will be set to now func",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo: "this is foo",
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				createdAt, updatedAt := vars[0], vars[2]
+
+				createdAtTime, ok := createdAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, createdAtTime.After(testStartedAt))
+
+				updatedAtTime, ok := updatedAt.(time.Time)
+				assert.True(t, ok)
+				assert.True(t, updatedAtTime.After(testStartedAt))
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			scope, err := scopeFromObjects(gdb, tc.slice, InsertFunc)
+
+			require.NotNil(t, scope)
+			require.NoError(t, err)
+
+			assert.Equal(t, tc.expectedSQL, scope.SQL)
+
+			tc.validFunc(t, scope.SQLVars)
+		})
+	}
+}

--- a/gorm_bulk_test.go
+++ b/gorm_bulk_test.go
@@ -515,6 +515,51 @@ func Test_Timestamps(t *testing.T) {
 				assert.True(t, updatedAtTime.After(testStartedAt))
 			},
 		},
+		{
+			description: "all records with empty time will be set to the same value",
+			slice: []interface{}{
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo: "this is foo",
+				},
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo: "this is bar",
+				},
+				struct {
+					Foo       string
+					CreatedAt time.Time
+					UpdatedAt time.Time
+				}{
+					Foo: "this is baz",
+				},
+			},
+			expectedSQL: "INSERT INTO `` (`created_at`, `foo`, `updated_at`) VALUES (?, ?, ?), (?, ?, ?), (?, ?, ?)",
+			validFunc: func(t *testing.T, vars []interface{}) {
+				var bulkTime time.Time
+				for i, columnNum := 0, 3; i < len(vars)/columnNum; i++ {
+					createdAt, updatedAt := vars[i*columnNum+0], vars[i*columnNum+2]
+					if bulkTime.IsZero() {
+						bulkTime = createdAt.(time.Time)
+						assert.True(t, bulkTime.After(testStartedAt))
+					}
+
+					createdAtTime, ok := createdAt.(time.Time)
+					assert.True(t, ok)
+					assert.True(t, createdAtTime.Equal(bulkTime))
+
+					updatedAtTime, ok := updatedAt.(time.Time)
+					assert.True(t, ok)
+					assert.True(t, updatedAtTime.Equal(bulkTime))
+				}
+			},
+		},
 	}
 
 	for _, tc := range cases {

